maps/time_key_map.go

```diff
@@ -0,0 +1,163 @@
+package maps
+
+import (
+	"time"
+
+	"github.com/akitasoftware/go-utils/sets"
+)
+
+func getInternalMapKey(t time.Time) int64 {
+	return t.Unix()
+}
+
+func getReverseKey(key int64) time.Time {
+	return time.Unix(key, 0)
+}
+
+type TimeMap[V any] struct {
+	internalMap map[int64]V
+}
+
+func NewTimeMap[V any]() TimeMap[V] {
+	return TimeMap[V]{
+		internalMap: make(map[int64]V),
+	}
+}
+
+func (tm TimeMap[V]) Put(k time.Time, v V) {
+	key := getInternalMapKey(k)
+	tm.internalMap[key] = v
+}
+
+func (m TimeMap[V]) Upsert(k time.Time, v V, onConflict func(v, newV V) V) {
+	newV := v
+	key := getInternalMapKey(k)
+	if oldV, exists := m.internalMap[key]; exists {
+		newV = onConflict(oldV, newV)
+	}
+	m.internalMap[key] = newV
+}
+
+// If the key k is not already in the map, then it is entered into the map with
+// the result of calling the supplied function. If the function returns an
+// error, then the map is not modified, and the error is returned.
+func (m TimeMap[V]) ComputeIfAbsent(k time.Time, computeValue func() (V, error)) error {
+	_, err := m.GetOrCompute(k, computeValue)
+	return err
+}
+
+// If the key k is not already in the map, then it is entered into the map with
+// the result of calling the supplied function.
+func (m TimeMap[V]) ComputeIfAbsentNoError(k time.Time, computeValue func() V) {
+	m.GetOrComputeNoError(k, computeValue)
+}
+
+func (m TimeMap[V]) Add(other TimeMap[V], onConflict func(v, newV V) V) {
+	for k, v := range other.internalMap {
+		timeKey := getReverseKey(k)
+		m.Upsert(timeKey, v, onConflict)
+	}
+}
+
+func (tm TimeMap[V]) Get(k time.Time) (V, bool) {
```

Comment:
**Blocking.** Let's make this consistent with `Map.Get` and return an `optionals.Optional[V]`.